scripts/deployment/deploy.py

```diff
@@ -0,0 +1,182 @@
+        database=conn_config.database,
+        user=conn_config.username,
+        password=conn_config.password,
+        host=conn_config.host,
+        port=conn_config.port
+    )
+
+    conn.autocommit = True
+    cursor = conn.cursor()
+
+    cursor.execute(sql)
+
+    conn.commit()
+    conn.close()
+
+
+def read_file(file_name: str) -> str:
+    logging.debug(f"    - reading file `{file_name}`")
+    file = open(file_name, "r")
+    result = file.read()
+    file.close()
+    return result
+
+
+def ensure_trailing_slash(path: str) -> str:
+    if path.endswith('/'):
+        return path
+    else:
+        return path + '/'
+
+
+def process_dir(directory: str, conn_config: PostgresDBConn, create_db: bool):
+    logging.info(f"Picking up source files from directory `{directory}`")
+    public_schema = "public"
+    root = next(os.walk(directory), (None, [], []))
+    schemas = list(root[1])
+    files = list(filter(lambda fn: fn.endswith(".ddl"), root[2]))
+
+    # process root files
+    database_creation_sqls = []
+    init_sqls = []
+    for filename in files:
+        if filename.startswith("00_"):
+            database_creation_sqls.append(filename)
+        else:
+            init_sqls.append(filename)
+    # process schemas
+    schemas_sqls = []
+    if public_schema in schemas:
+        # public folder has to go first
+        schemas.remove(public_schema)
+        schemas_sqls += process_schema(directory, public_schema, False)
+
+    for schema in schemas:
+        schemas_sqls += process_schema(directory, schema, True)
+
+    # execute the collected Sqls
+    if (len(database_creation_sqls) > 0) and create_db:
+        logging.info("Creating database")
+        db_conn_config = copy.copy(conn_config)
+        db_conn_config.database = "postgres"
+        sql = "\n".join(map(read_file, database_creation_sqls))
+        execute_sql(db_conn_config, sql)
+
+    if len(init_sqls) > 0:
+        logging.info("Initializing the database")
+        sql = "\n".join(map(read_file, init_sqls))
+        execute_sql(conn_config, sql)
+    if len(schemas_sqls) > 0:
+        logging.info("Populating the schemas")
+        sql = "\n".join(schemas_sqls)
+        execute_sql(conn_config, sql)
+
+
+def process_schema(base_dir: str,  schema_name: str, expect_schema_creation: bool) -> list[str]:
+    logging.info(f"  - schema '{schema_name}'")
+    schema_dir = ensure_trailing_slash(base_dir + schema_name)
+    schema_creation = []
+    functions = []
+    tables = []
+    has_schema_creation = False
+    files = os.listdir(schema_dir)
+    files.sort()
+    for input_file in files:
+        if input_file == "_.ddl":
+            schema_creation = [read_file(schema_dir + input_file)]
+            has_schema_creation = True
+        elif input_file.endswith(".sql"):
+            functions.append(read_file(schema_dir + input_file))
+        elif input_file.endswith(".ddl"):
+            tables.append(read_file(schema_dir + input_file))
```

Comment:
potential optimization could be to only get the file names in this function, and then the loading of their content could be done in the same way as on line `"\n".join(map(read_file, init_sqls))`, so it would be consistent and a bit more memory efficient

Reply:
More consitent maybe.
More memory efficient why?

Reply:
> More memory efficient why?

because now it has to be hold in memory from this point until it get's garbage collected which is probably after some point of running the sql statements.

If read only in that join function, then the memory requirements are the same but it will be held in memory probably for shorter time. Not a big deal in this scenario I know